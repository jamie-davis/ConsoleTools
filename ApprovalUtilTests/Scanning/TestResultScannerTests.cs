using ApprovalUtil.Scanning;
using ApprovalUtilTests.TestUtilities;
using TestConsoleLib;
using TestConsoleLib.Testing;

namespace ApprovalUtilTests.Scanning;

public class TestResultScannerTests
{
    [Fact]
    public void ResultsAreExtracted()
    {
        //Arrange
        using var testData = new TestOutputGenerator();
        testData.MakeTest("TestClass1", "TestOne");
        testData.MakeTest("TestClass1", "TestTwo", true);
        testData.MakeTest("TestClass1", "TestThree");
        testData.MakeTest("TestClass2", "TestOne", true);
        testData.MakeTest("TestClass2", "TestFailed", createFailed: true);
        testData.MakeTest("TestClass2", "MixedLineEndingsPass", true, mixLineEndings: true);

        //Act
        var result = TestResultScanner.Scan(testData.FolderPath);

        //Assert
        var output = new Output();
        output.WrapLine(@$"""Passed"" indicates the result returned by {nameof(TestResultScanner)}. ""File Exact Match"" indicates whether the received file exactly matches the approved file on disk.");
        output.WriteLine();
        output.WrapLine("The scanner's comparison is line-ending agnostic, so it will report a match even if the line endings are different. One of the test files has deliberately been written with line ending differences to test this, and should show up as a pass but not an exact match.");
        output.WriteLine();
        var formattedResult = result
            .Select(a => new
            {
                a.Test.TestTypeName,
                a.Test.TestName,
                Approved = Path.GetFileName(a.Test.ApprovedFile),
                a.Passed,
                FileExactMatch = File.Exists(a.Test.ReceivedFile) ? (bool?)(File.ReadAllText(a.Test.ApprovedFile) == File.ReadAllText(a.Test.ReceivedFile)) : null 
            })
            .OrderBy(a => a.TestTypeName)
            .ThenBy(a => a.TestName);
        output.FormatTable(formattedResult);
        output.Report.Verify();
    }
}