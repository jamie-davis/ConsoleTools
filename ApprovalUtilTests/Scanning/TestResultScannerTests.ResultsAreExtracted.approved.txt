"Passed" indicates the result returned by TestResultScanner. "File Exact Match" indicates      
whether the received file exactly matches the approved file on disk.                           
                                                                                               
The scanner's comparison is line-ending agnostic, so it will report a match even if the line   
endings are different. One of the test files has deliberately been written with line ending    
differences to test this, and should show up as a pass but not an exact match.                 
                                                                                               
Test Type                                                                           File Exact 
Name       Test Name            Approved                                     Passed Match      
---------- -------------------- -------------------------------------------- ------ -----------
TestClass1 TestOne              TestClass1.TestOne.approved.txt              True              
TestClass1 TestThree            TestClass1.TestThree.approved.txt            True              
TestClass1 TestTwo              TestClass1.TestTwo.approved.txt              True   True       
TestClass2 MixedLineEndingsPass TestClass2.MixedLineEndingsPass.approved.txt True   False      
TestClass2 TestFailed           TestClass2.TestFailed.approved.txt           False  False      
TestClass2 TestOne              TestClass2.TestOne.approved.txt              True   True       
                                                                                               